server/src/main/java/org/elasticsearch/common/file/AbstractFileWatchingService.java

```diff
@@ -237,14 +238,16 @@ protected final void watcherThread() {
                 key.reset();
 
                 if (key == settingsDirWatchKey) {
-                    // there may be multiple events for the same file - we only want to re-read once
-                    Set<Path> processedFiles = new HashSet<>();
-                    for (WatchEvent<?> e : events) {
-                        Path fullFile = settingsDir.resolve(e.context().toString());
-                        if (processedFiles.add(fullFile)) {
-                            if (fileChanged(fullFile)) {
-                                process(fullFile);
-                            }
+                    Set<Path> changedPaths = events.stream()
+                        .map(event -> settingsDir.resolve(event.context().toString()))
+                        .collect(Collectors.toSet());
+                    for (var changedPath : changedPaths) {
+                        // If a symlinked dir changed in the settings dir, it could be linked to other symlinks, so reprocess all files
+                        if (Files.isDirectory(changedPath) && Files.isSymbolicLink(changedPath)) {
```

Comment:
I believe isDirectory and isSymbolic link need to be extracted to helper methods the way other Files methods are, so that these operations can be granted via entitlements for modules outside server.

Reply:
I've moved the file operations to their own methods. I'm trying to understand why this works with the `MultiProjectFileSettingsService` and I think it might be because `getDeclaringClass` is used to determine the caller. In this case the caller is `MultiProjectFileSettingsService` but since we don't override the methods in `FileSettingsService` the calling class becomes `FileSettingsService` since that's where the method is declared. `FileSettingsService` in turn is part of server, so that's why this works?

Reply:
Without overriding it, the calling class appears to be in the server module, so server policies will apply. Extracting methods and overriding them in the module is the _current_ way to make the call appear as originated from the module, so the module's policy appears. 
It could make no difference today e.g. all modules are granted read access to the config directory, but that might change, so better to structure the code so it resolves to the correct module/policy

Reply:
(we are designing ways to make this easier and more explicit in the future, but atm this is the way with entitlements)

Reply:
Thanks for explaining @ldematte! 

For the `serverless-multi-project` module, is the entitlement-policy inherited from the extended plugin (core in this case) and is that why it works without an explicit policy? I've update the `MultiProjectFileSettingsService` to override the files methods in https://github.com/elastic/elasticsearch-serverless/pull/3881